infrastructure/src/main/java/org/corfudb/infrastructure/BatchProcessor.java

```diff
@@ -0,0 +1,216 @@
+import org.corfudb.infrastructure.log.StreamLog;
+import org.corfudb.protocols.wireprotocol.CorfuPayloadMsg;
+import org.corfudb.protocols.wireprotocol.LogData;
+import org.corfudb.protocols.wireprotocol.RangeWriteMsg;
+import org.corfudb.protocols.wireprotocol.TailsResponse;
+import org.corfudb.protocols.wireprotocol.TrimRequest;
+import org.corfudb.protocols.wireprotocol.WriteRequest;
+import org.corfudb.runtime.exceptions.WrongEpochException;
+import org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuInterruptedError;
+
+/**
+ * This class manages access for operations that need ordering while executing against
+ * the backing storage.
+ */
+@Slf4j
+public class BatchProcessor implements AutoCloseable {
+
+    static private final int BATCH_SIZE = 50;
+
+    final private boolean sync;
+
+    private StreamLog streamLog;
+
+    final private BlockingQueue<BatchWriterOperation> operationsQueue;
+
+    private ExecutorService processorService = Executors
+            .newSingleThreadExecutor(new ThreadFactoryBuilder()
+                    .setDaemon(false)
+                    .setNameFormat("LogUnit-BatchProcessor-%d")
+                    .build());
+
+    /**
+     * The sealEpoch is the epoch up to which all operations have been sealed. Any
+     * BatchWriterOperation arriving after the sealEpoch with an epoch less than the sealEpoch
+     * is completed exceptionally with a WrongEpochException.
+     * This is persisted in the ServerContext by the LogUnitServer to withstand restarts.
+     */
+    private long sealEpoch;
+
+    /**
+     * Returns a new BatchProcessor for a stream log.
+     *
+     * @param streamLog      stream log for writes (can be in memory or file)
+     * @param sealEpoch All operations stamped with epoch less than the epochWaterMark are
+     *                       discarded.
+     * @param streamLog stream log for writes (can be in memory or file)
+     * @param sync    If true, the batch writer will sync writes to secondary storage
+     */
+    public BatchProcessor(StreamLog streamLog, long sealEpoch, boolean sync) {
+        this.sealEpoch = sealEpoch;
+        this.sync = sync;
+        this.streamLog = streamLog;
+        operationsQueue = new LinkedBlockingQueue<>();
+        processorService.submit(this::processor);
+    }
+
+    /**
+     * Add a task to the processor.
+     * @param type The request type
+     * @param msg The request message
+     * @return returns a future result for the request, if it expects one
+     */
+    public <T> CompletableFuture <T> addTask(@Nonnull Type type, @Nonnull CorfuPayloadMsg msg) {
+        BatchWriterOperation<T> operation = new BatchWriterOperation<>(type, msg);
+        operationsQueue.add(operation);
+        return operation.getFutureResult();
+    }
+
+    private void processor() {
+
+        if (!sync) {
+            log.warn("batchWriteProcessor: writes configured to not sync with secondary storage");
```

Comment:
Should we add some info on the implications, so it is more informative to users. Could incur in data loss?

Reply:
I think it would be redundant. Its kind of implied with "configured to not sync with secondary storage"